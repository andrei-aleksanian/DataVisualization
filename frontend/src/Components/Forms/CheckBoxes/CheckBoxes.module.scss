.index {
  margin: 1rem 0;
}

.index p {
  font-size: 1.1rem;
}

.CheckBoxes {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.CheckBoxes label {
  display: block;
  position: relative;
  cursor: pointer;
  margin: 1rem 0;
}

.CheckBoxes label:nth-child(1) {
  margin: 1rem 2rem 0 0;
}

.CheckBoxes label input {
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;
}

// Custom checkbox
.checkmark {
  position: absolute;
  top: 0;
  left: 0;
  height: 15px;
  width: 15px;
  background-color: var(--color-text);
  border: 1px solid var(--color-action);
  border-radius: 3px;
}

.CheckBoxes label input:checked ~ .checkmark {
  background-color: var(--color-action);
}

.text {
  margin: 0 0 0 2rem;
}
