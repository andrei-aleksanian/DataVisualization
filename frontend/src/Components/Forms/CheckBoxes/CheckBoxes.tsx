/**
 * Checkboxes - radio buttons.
 *
 * Used for displaying an array of checkboxes
 * that bind to state of the parent component.
 *
 * @param entries - array of checkboxes to be generated
 * @param currentValue - current value of the state from parent component
 * @param onChange - function that handles a ckick on any checkbox
 * @param labelText - text to describe the group of checkboxes. e.g. 'Checkboxes:'
 */

import getId from 'utils/getId';
import Label from '../Label';
import classes from './CheckBoxes.module.scss';

export interface CheckBoxesProps {
  entries: { value: number; text: string }[];
  currentValue: number;
  onChange: (event: React.ChangeEvent, currentValue: number) => void;
  labelText: string;
  tooltipText: string;
}

const CheckBoxes = ({ currentValue, onChange, entries, labelText, tooltipText }: CheckBoxesProps) => (
  <div className={classes.index}>
    <Label text={labelText} tooltipText={tooltipText} />
    <div className={classes.CheckBoxes}>
      {entries.map((e) => (
        <label key={getId('checkbox')}>
          <input type="radio" checked={currentValue === e.value} onChange={(event) => onChange(event, e.value)} />
          <span className={classes.text}>{e.text}</span>
          <span className={classes.checkmark} />
        </label>
      ))}
    </div>
  </div>
);

export default CheckBoxes;
