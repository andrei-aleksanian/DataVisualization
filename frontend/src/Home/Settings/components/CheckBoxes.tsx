import { Algorithm } from 'types/Settings';
import classes from './CheckBoxes.module.scss';

export interface CheckBoxesProps {
  currentAlgorithm: Algorithm;
  onClickAlgorithm: (event: React.ChangeEvent, algorithm: Algorithm) => void;
}

const CheckBoxes = ({ currentAlgorithm, onClickAlgorithm }: CheckBoxesProps) => (
  <div className={classes.index}>
    <p>Algorithm:</p>
    <div className={classes.CheckBoxes}>
      <label>
        <input
          type="radio"
          checked={currentAlgorithm === Algorithm.COVA}
          onChange={(e) => onClickAlgorithm(e, Algorithm.COVA)}
        />
        <span className={classes.text}>COVA</span>
        <span className={classes.checkmark} />
      </label>
      <label>
        <input
          type="radio"
          checked={currentAlgorithm === Algorithm.ANGEL}
          onChange={(e) => onClickAlgorithm(e, Algorithm.ANGEL)}
        />
        <span className={classes.text}>ANGEL</span>
        <span className={classes.checkmark} />
      </label>
    </div>
  </div>
);

export default CheckBoxes;
