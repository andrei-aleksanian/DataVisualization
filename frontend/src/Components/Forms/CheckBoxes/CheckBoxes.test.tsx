import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CheckBoxes, { CheckBoxesProps } from './CheckBoxes';

enum TestEntries {
  BOX1,
  BOX2,
}
let state = TestEntries.BOX1;
const setState = (event: React.ChangeEvent, currentValue: number) => {
  event.preventDefault();
  state = currentValue;
};
const testCheckBoxesProps: CheckBoxesProps = {
  entries: [
    { value: TestEntries.BOX1, text: 'box1' },
    { value: TestEntries.BOX2, text: 'box2' },
  ],
  currentValue: state,
  onChange: setState,
  labelText: 'Test Checkboxes',
};

const cleanupState = () => {
  state = TestEntries.BOX1;
};

const init = () => {
  render(<CheckBoxes {...testCheckBoxesProps} />);
};

describe('Test UI renders', () => {
  test('Settings matches snapshot', async () => {
    const { asFragment } = render(<CheckBoxes {...testCheckBoxesProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Heading is rendered', () => {
    init();
    expect(screen.getByText(testCheckBoxesProps.labelText)).toBeInTheDocument();
  });

  test('Test all entries from entries array render', () => {
    init();
    for (let i = 0; i < testCheckBoxesProps.entries.length; i += 1) {
      const radio = screen.getByRole('radio', { name: testCheckBoxesProps.entries[i].text });
      expect(radio).toBeInTheDocument();
    }
  });
});

describe('Test UI changes and events', () => {
  test('Test click on a checkbox changes outside state', () => {
    init();
    const checkBox = screen.getByRole('radio', { name: testCheckBoxesProps.entries[1].text });
    userEvent.click(checkBox);

    expect(state).toBe(TestEntries.BOX2);
    cleanupState();
  });
});
