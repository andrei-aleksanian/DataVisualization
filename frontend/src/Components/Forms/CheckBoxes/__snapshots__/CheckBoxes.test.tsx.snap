// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test UI renders Settings matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="index"
  >
    <div
      class="index"
    >
      <p>
        Test Checkboxes
      </p>
      <div
        class="Tooltip"
      >
        <img
          alt="info-0"
          src="info.svg"
        />
        <div
          style="top: 0px; left: 30px;"
        >
          Choose how many points should the algorithm assume to have for neighbour separation
        </div>
      </div>
    </div>
    <div
      class="CheckBoxes"
    >
      <label>
        <input
          checked=""
          type="radio"
        />
        <span
          class="text"
        >
          box1
        </span>
        <span
          class="checkmark"
        />
      </label>
      <label>
        <input
          type="radio"
        />
        <span
          class="text"
        >
          box2
        </span>
        <span
          class="checkmark"
        />
      </label>
    </div>
  </div>
</DocumentFragment>
`;
